### PR TITLE
support xen-pv

### PR DESCRIPTION
**Pull request by gao-feng**
Env:
xen: 4.9
qemu: 2.10.0
guest kernel: 4.12.4 

Todo:
a: Use one xenlight library for xen pv and hvm.
b: use vm_console to handle pv domain console output (done)
c: fix some weird problems in some wrong case.
1, host dom name will be (null) and fail to create xenpv domain(fixed)
2,hyperd panic when fail to create xenpv domain (fixed)
3,hyperd hang (disappear after fix bug 2)

Signed-off-by: Gao feng <omarapazanadi@gmail.com>

### CODE REVIEW

File: hypervisor/xenpv/xenpv.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+	xc.connectConsole(ctx.Id, xc.domId)
+	go func() {
+		//unpause dom until runv connected to the socket, otherwise may loss the ready message
+		ctx.WaitSockConnected()
+		xctx.DomainUnpause(xc.domId)
+	}()
+}
+
+func (xc *XenPvContext) connectConsole(vmid string, domId xl.Domid) error {
```
Comment (laijs) [inline hypervisor/xenpv/xenpv.go]: Could it be possible to merge it into vm_console.go?

Comment (gao-feng) [inline hypervisor/xenpv/xenpv.go]: I can only add another unix socket to pass console data, seems like pv doesn't have direct way to get console from unix socket.

Comment (laijs) [inline hypervisor/xenpv/xenpv.go]: It seems that adding `ConnectConsole() io.Reader`  to the driver interface is a possible way to merge the code. What do you think?

Comment (gao-feng) [inline hypervisor/xenpv/xenpv.go]: @laijs 
Good point!
Commit two adds ConsoleReader to the driver interface,  please check.

File: hypervisor/vm_console.go [hunk 2]
```diff
@@ -59,20 +59,19 @@ func watchVmConsole(ctx *VmContext) {
...
-		ctx.Log(ERROR, "failed to connected to %s: %v", ctx.ConsoleSockName, err)
+		ctx.Log(ERROR, "failed to get console for %v %v", ctx.Id, err)
 		return
 	}
+	defer reader.Close()
 
-	ctx.Log(TRACE, "connected to %s", ctx.ConsoleSockName)
-
-	tc, err := telnet.NewConn(conn)
+	tc, err := telnet.NewConn(reader)
```
Comment (laijs) [inline hypervisor/vm_console.go]: It seams that the above code(L62-L69) should all be moved in to DCtx.ConsoleReader() (those who are using ctx.ConsoleSockName).
DCtx.ConsoleReader() for xenpv doesn't returns a telnet socket, you cannot use telnet.NewConn() for it.

**Comment by gao-feng** [comment]
@laijs 
updated, xenlight-runv.go needs to access the private data `Context.ctx` in xenlight package, looks like use xenlight package as vendor is impossible.

**Comment by gnawux** [comment]
hmm.... looks we need submit change request to official xenlight go binding in the future

**Pull request merged**